mbed_project/_internal/git_utils.py

```diff
@@ -0,0 +1,77 @@
+#
+# Copyright (C) 2020 Arm Mbed. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+"""Wrappers for git operations."""
+from dataclasses import dataclass
+from pathlib import Path
+
+import git
+
+from mbed_project.exceptions import VersionControlError
+from mbed_project._internal.progress import ProgressReporter
+
+
+@dataclass
+class GitReference:
+    """Git reference for a remote repository.
+
+    Attributes:
+        repo_url: URL of the git repository.
+        ref: The reference commit sha, tag or branch.
+    """
+
+    repo_url: str
+    ref: str
+
+
+def clone(url: str, dst_dir: Path) -> git.Repo:
+    """Clone a library repository.
+
+    Args:
+        git_ref: Object containing git information.
+        dst_dir: Destination directory for the cloned repo.
+
+    Raises:
+        VersionControlError: Cloning the repository failed.
+    """
+    try:
+        return git.Repo.clone_from(url, str(dst_dir), progress=ProgressReporter(name=url))
+    except git.exc.GitCommandError as err:
+        raise VersionControlError(f"Cloning git repository from url '{url}' failed. Error from VCS: {err.stderr}")
+
+
+def checkout(repo: git.Repo, ref: str) -> None:
+    """Check out the at specific reference in the given repository.
+
+    Args:
+        repo: git.Repo object where the checkout will be performed.
+        ref: Git commit hash, branch or tag reference.
+
+    Raises:
+        VersionControlError: Check out failed.
+    """
+    git_object = git.repo.fun.name_to_object(repo, ref)
```

Comment:
what happens if the `ref` does not exist? would it only fail during checkout?

Reply:
shouldn't `self.ref` be populated after this?

Reply:
`ref` should always exist here, it's checked outside of this function

Reply:
That's handled by the library handling code, these functions just wrap the git commands.

Reply:
ok. May be worth putting a comment there in the argument description.

Reply:
I think `mypy` saves writing a comment as the type would be `Optional[str]` if it could be `None`, right?

Reply:
I was more thinking a comment saying that the ref must be valid as it must be a hash, tag or branch defined in the repo

Reply:
Ah, I see what you mean. Yep, will do that. 👍

Reply:
b5f35bcacdd6364690149de0e09736ba03804046